worker/presence/presence_test.go

```diff
@@ -0,0 +1,407 @@
+		Model:           "model-uuid",
+		Server:          "machine-5",
+		Agent:           "agent-2",
+		ConnectionID:    42,
+		Status:          corepresence.Alive,
+		ControllerAgent: true,
+		UserData:        "test",
+	})
+}
+
+func (s *PresenceSuite) TestDisconnectTopic(c *gc.C) {
+	w := s.worker(c)
+	defer workertest.CleanKill(c, w)
+
+	connect(s.recorder, agent1, agent2)
+
+	done, err := s.hub.Publish(
+		apiserver.DisconnectTopic,
+		apiserver.APIConnection{
+			Origin:       agent2.Server,
+			ConnectionID: agent2.ConnectionID,
+		})
+	c.Assert(err, jc.ErrorIsNil)
+	s.AssertDone(c, done)
+	s.AssertConnections(c, alive(agent1))
+}
+
+func (s *PresenceSuite) TestPresenceRequest(c *gc.C) {
+	w := s.worker(c)
+	defer workertest.CleanKill(c, w)
+
+	connect(s.recorder, agent1, agent2, agent3, agent4)
+
+	done := make(chan struct{})
+	unsub, err := s.hub.Subscribe(apiserver.PresenceResponseTopic, func(topic string, data apiserver.PresenceResponse, err error) {
+		c.Logf("handler called for %q", topic)
+		c.Check(err, jc.ErrorIsNil)
+		c.Check(data.Origin, gc.Equals, ourServer)
+
+		c.Check(data.Connections, gc.HasLen, 2)
+		s.CheckConnection(c, data.Connections[0], agent1)
+		s.CheckConnection(c, data.Connections[1], agent3)
+
+		close(done)
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	defer unsub()
+
+	// When connections are established from them to us, we ask for their presence info.
+	_, err = s.hub.Publish(
+		apiserver.PresenceRequestTopic,
+		apiserver.OriginTarget{Origin: otherServer, Target: ourServer})
+	c.Assert(err, jc.ErrorIsNil)
+	s.AssertDone(c, done)
+}
+
+func (s *PresenceSuite) TestPresenceRequestOtherServer(c *gc.C) {
+	w := s.worker(c)
+	defer workertest.CleanKill(c, w)
+
+	called := make(chan struct{})
+	unsub, err := s.hub.Subscribe(apiserver.PresenceResponseTopic, func(topic string, data apiserver.PresenceResponse, err error) {
+		c.Logf("handler called for %q", topic)
+		close(called)
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	defer unsub()
+
+	// When connections are established from them to us, we ask for their presence info.
+	_, err = s.hub.Publish(
+		apiserver.PresenceRequestTopic,
+		apiserver.OriginTarget{Origin: otherServer, Target: "another"})
+	c.Assert(err, jc.ErrorIsNil)
+	s.AssertNotCalled(c, called)
+}
+
+func (s *PresenceSuite) TestPresenceResponse(c *gc.C) {
+	w := s.worker(c)
+	defer workertest.CleanKill(c, w)
+
+	connect(s.recorder, agent1, agent2, agent3, agent4)
+	s.recorder.ServerDown(otherServer)
+
+	// When connections are established from them to us, we ask for their presence info.
+	done, err := s.hub.Publish(
+		apiserver.PresenceResponseTopic,
+		apiserver.PresenceResponse{
+			Origin: otherServer,
+			Connections: []apiserver.APIConnection{
+				apiConn(agent2), apiConn(agent4),
+			},
+		})
+	c.Assert(err, jc.ErrorIsNil)
+	s.AssertDone(c, done)
+
+	s.AssertConnections(c, alive(agent1), alive(agent2), alive(agent3), alive(agent4))
+}
+
+func (s *PresenceSuite) AssertDone(c *gc.C, called <-chan struct{}) {
```

Comment:
these helpers don't need to be exported

Reply:
I know, but the capitalisation of the method names fits nicely with my OCD and having them line up with other Assert functions.